include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -456,96 +457,75 @@ __parallel_transform_reduce(oneapi::dpl::__internal::__device_backend_tag __back
-                __backend_tag, ::std::forward<_ExecutionPolicy>(__exec), __n, __reduce_op, __transform_op, __init,
-                ::std::forward<_Ranges>(__rngs)...);
-        }
-        else if (__n <= 8192)
-        {
-            return __parallel_transform_reduce_small_impl<_Tp, 256, 32, _Commutative>(
-                __backend_tag, ::std::forward<_ExecutionPolicy>(__exec), __n, __reduce_op, __transform_op, __init,
-                ::std::forward<_Ranges>(__rngs)...);
-        }
+    // Limit work-group size to __max_work_group_size for performance on GPUs. Empirically tested.
+    __work_group_size = std::min(__work_group_size, __max_work_group_size);
+    const _Size __max_elements_per_wg = __work_group_size * __max_iters_per_work_item;
 
-        // Use two-step tree reduction.
-        // First step reduces __work_group_size * __iters_per_work_item_device_kernel elements.
-        // Second step reduces __work_group_size * __iters_per_work_item_work_group_kernel elements.
-        else if (__n <= 2097152)
-        {
-            return __parallel_transform_reduce_mid_impl<_Tp, 256, 32, 1, _Commutative>(
-                __backend_tag, ::std::forward<_ExecutionPolicy>(__exec), __n, __reduce_op, __transform_op, __init,
-                ::std::forward<_Ranges>(__rngs)...);
-        }
-        else if (__n <= 4194304)
-        {
-            return __parallel_transform_reduce_mid_impl<_Tp, 256, 32, 2, _Commutative>(
-                __backend_tag, ::std::forward<_ExecutionPolicy>(__exec), __n, __reduce_op, __transform_op, __init,
-                ::std::forward<_Ranges>(__rngs)...);
-        }
-        else if (__n <= 8388608)
-        {
-            return __parallel_transform_reduce_mid_impl<_Tp, 256, 32, 4, _Commutative>(
-                __backend_tag, ::std::forward<_ExecutionPolicy>(__exec), __n, __reduce_op, __transform_op, __init,
-                ::std::forward<_Ranges>(__rngs)...);
-        }
-        else if (__n <= 16777216)
-        {
-            return __parallel_transform_reduce_mid_impl<_Tp, 256, 32, 8, _Commutative>(
-                __backend_tag, ::std::forward<_ExecutionPolicy>(__exec), __n, __reduce_op, __transform_op, __init,
-                ::std::forward<_Ranges>(__rngs)...);
-        }
-        else if (__n <= 33554432)
-        {
-            return __parallel_transform_reduce_mid_impl<_Tp, 256, 32, 16, _Commutative>(
-                __backend_tag, ::std::forward<_ExecutionPolicy>(__exec), __n, __reduce_op, __transform_op, __init,
-                ::std::forward<_Ranges>(__rngs)...);
-        }
-        else if (__n <= 67108864)
+    // Use single work group implementation if less than __max_iters_per_work_item elements per work-group.
+    // We can use 16-bit addressing since we have at most __max_work_group_size * __max_iters_per_work_item elements.
+    if (__n <= __max_elements_per_wg)
+    {
+        const auto __n_short = static_cast<std::uint16_t>(__n);
```

Comment:
While I don't disagree with this statement in the comments, it is dependent on the numbers above.  Can we do more to protect from future bugs?  
A couple options:
1) Add an assert that `__max_work_group_size * __max_iters_per_work_item`  is less than numerical limits of uint16_t. 
2) Add a comment to the empirically found tuning parameters, mentioning that index types might need to be adjusted if there are any changes to these values.

Reply:
I've added a comment and an assert.

Reply:
I think a `static_assert` of the "max" constants vs the numerical max is the better option, because with this assert we wont know about the issue until a person runs an example in debug mode which has more elements than the type can hold, which may not be at the same time as the tuning parameters are tweaked.
  
If we use a static assert, we can know as soon as we compile new tuning parameters which breaks our type rules.

Reply:
Good point. I've changed it.